File: cli/scr.js
#!/usr/bin/env node
'use strict';

const http = require('http');
const https = require('https');
const fs = require('fs');
const path = require('path');
const os = require('os');
const { URL } = require('url');

// ========== Config ==========
const CONFIG_PATH = path.join(os.homedir(), '.scr.json');

function loadConfig() {
  try {
    if (fs.existsSync(CONFIG_PATH)) return JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf8'));
  } catch (e) { /* ignore */ }
  return {};
}

function saveConfig(cfg) {
  fs.writeFileSync(CONFIG_PATH, JSON.stringify(cfg, null, 2), 'utf8');
}

// ========== Argument Parser ==========
function parseArgs(argv) {
  const args = { _: [] };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '-r' || a === '--room') { args.room = argv[++i]; }
    else if (a === '-p' || a === '--password') { args.password = argv[++i]; }
    else if (a === '-n' || a === '--name') { args.name = argv[++i]; }
    else if (a === '--server') { args.server = argv[++i]; }
    else if (a === '--limit') { args.limit = parseInt(argv[++i], 10); }
    else { args._.push(a); }
  }
  return args;
}

// ========== HTTP Helper ==========
function request(method, urlStr, body, headers) {
  return new Promise((resolve, reject) => {
    const u = new URL(urlStr);
    const mod = u.protocol === 'https:' ? https : http;
    const opts = {
      hostname: u.hostname,
      port: u.port || (u.protocol === 'https:' ? 443 : 80),
      path: u.pathname + u.search,
      method,
      headers: { ...headers }
    };
    let payload = null;
    if (body !== undefined && body !== null) {
      payload = typeof body === 'string' ? body : JSON.stringify(body);
      opts.headers['Content-Type'] = 'application/json';
      opts.headers['Content-Length'] = Buffer.byteLength(payload);
    }
    const req = mod.request(opts, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        try {
          resolve({ status: res.statusCode, data: JSON.parse(data) });
        } catch (e) {
          resolve({ status: res.statusCode, data: { ok: false, error: data || 'invalid response' } });
        }
      });
    });
    req.on('error', (e) => reject(e));
    if (payload) req.write(payload);
    req.end();
  });
}

// ========== Resolve Server URL ==========
function resolveServer(args) {
  let server = null;
  if (args.server) server = args.server;
  else if (process.env.SCR_SERVER) server = process.env.SCR_SERVER;
  else {
    const cfg = loadConfig();
    if (cfg.server) server = cfg.server;
  }
  if (!server) return 'http://localhost:3000';
  server = server.replace(/\/+$/, '');
  // auto-prepend http:// if no protocol specified
  if (!/^https?:\/\//i.test(server)) server = 'http://' + server;
  return server;
}

// ========== Commands ==========

async function cmdLogin(args) {
  const room = args.room;
  const user = args.name;
  if (!room) { process.stderr.write('Error: -r <room> required\n'); process.exit(1); }
  if (!user) { process.stderr.write('Error: -n <name> required\n'); process.exit(1); }

  const server = resolveServer(args);
  const body = { room, user };
  if (args.password) body.password = args.password;

  try {
    const res = await request('POST', server + '/api/login', body, {});
    if (res.data && res.data.ok) {
      saveConfig({
        server,
        token: res.data.token,
        room: res.data.room,
        roomName: res.data.roomName,
        user: res.data.user
      });
      process.stdout.write(`Logged in as ${res.data.user} to ${res.data.roomName} on ${server}\n`);
    } else {
      process.stderr.write(`Error: ${(res.data && res.data.error) || 'login failed'}\n`);
      process.exit(1);
    }
  } catch (e) {
    process.stderr.write(`Error: Cannot connect to server at ${server} - ${e.code || e.message || e}\n`);
    process.exit(1);
  }
}

async function cmdSend(args) {
  const cfg = loadConfig();
  if (!cfg.token) { process.stderr.write('Not logged in. Run: node cli/scr.js login -r <room> -n <name>\n'); process.exit(1); }

  const text = args._.join(' ');
  if (!text) { process.stderr.write('Error: message text required\n'); process.exit(1); }

  const server = resolveServer(args) || cfg.server;
  try {
    const res = await request('POST', server + '/api/send', { text }, {
      'Authorization': 'Bearer ' + cfg.token
    });
    if (res.data && res.data.ok) {
      process.stdout.write(`ok ${res.data.message.id}\n`);
    } else if (res.status === 401) {
      process.stderr.write('Session expired. Please login again.\n');
      process.exit(1);
    } else {
      process.stderr.write(`Error: ${(res.data && res.data.error) || 'send failed'}\n`);
      process.exit(1);
    }
  } catch (e) {
    process.stderr.write(`Error: Cannot connect to server at ${server} - ${e.code || e.message || e}\n`);
    process.exit(1);
  }
}

async function cmdRead(args) {
  const cfg = loadConfig();
  if (!cfg.token) { process.stderr.write('Not logged in. Run: node cli/scr.js login -r <room> -n <name>\n'); process.exit(1); }

  const afterId = args._[0] || null;
  const limit = args.limit || 50;
  const server = resolveServer(args) || cfg.server;

  let url = server + '/messages?room=' + encodeURIComponent(cfg.room) + '&limit=' + limit;
  if (afterId) url += '&after=' + encodeURIComponent(afterId);

  try {
    const res = await request('GET', url, null, {});
    if (res.data && res.data.ok && Array.isArray(res.data.messages)) {
      const msgs = res.data.messages;
      if (msgs.length === 0) {
        process.stdout.write('(no new messages)\n');
        return;
      }
      for (const m of msgs) {
        const ts = new Date(m.ts).toLocaleTimeString();
        const text = m.type === 'file' ? `[file] ${m.file ? m.file.name : ''}` : m.text;
        process.stdout.write(`[${m.id}] [${ts}] <${m.user}> ${text}\n`);
      }
      process.stdout.write(`LAST_ID:${msgs[msgs.length - 1].id}\n`);
    } else {
      process.stderr.write(`Error: ${(res.data && res.data.error) || 'read failed'}\n`);
      process.exit(1);
    }
  } catch (e) {
    process.stderr.write(`Error: Cannot connect to server at ${server} - ${e.code || e.message || e}\n`);
    process.exit(1);
  }
}

function cmdLogout() {
  if (fs.existsSync(CONFIG_PATH)) {
    fs.unlinkSync(CONFIG_PATH);
    process.stdout.write('Logged out. Session cleared.\n');
  } else {
    process.stdout.write('Not logged in.\n');
  }
}

function showHelp() {
  process.stdout.write(`scr - SimpleChatRoom CLI

Usage:
  scr login -r <room> -n <name> [-p <password>] [--server <url>]
  scr send <message...>
  scr read [lastMsgId] [--limit N]
  scr logout

Options:
  -r, --room      Room name
  -n, --name      Username
  -p, --password  Room password (if required)
  --server        Server URL (default: http://localhost:3000)
  --limit         Max messages to read (default: 50)

Config: ~/.scr.json
Server priority: --server > $SCR_SERVER > config > localhost:3000
`);
}

// ========== Main ==========
async function main() {
  const argv = process.argv.slice(2);
  if (argv.length === 0) { showHelp(); return; }

  const cmd = argv[0];
  const args = parseArgs(argv.slice(1));

  switch (cmd) {
    case 'login': await cmdLogin(args); break;
    case 'logout': cmdLogout(); break;
    case 'send': await cmdSend(args); break;
    case 'read': await cmdRead(args); break;
    case 'help': case '--help': case '-h': showHelp(); break;
    default:
      process.stderr.write(`Unknown command: ${cmd}\nRun: scr help\n`);
      process.exit(1);
  }
}

main().catch(e => {
  process.stderr.write(`Fatal: ${e.message}\n`);
  process.exit(1);
});

File: cli/scr.py
#!/usr/bin/env python3
"""scr - SimpleChatRoom CLI"""

import sys
import os
import json
import urllib.request
import urllib.parse
import urllib.error
from datetime import datetime

CONFIG_PATH = os.path.join(os.path.expanduser("~"), ".scr.json")


# ========== Config ==========

def load_config():
    try:
        if os.path.exists(CONFIG_PATH):
            with open(CONFIG_PATH, "r", encoding="utf-8") as f:
                return json.load(f)
    except Exception:
        pass
    return {}


def save_config(cfg):
    with open(CONFIG_PATH, "w", encoding="utf-8") as f:
        json.dump(cfg, f, indent=2, ensure_ascii=False)


# ========== Argument Parser ==========

def parse_args(argv):
    args = {"_": []}
    i = 0
    while i < len(argv):
        a = argv[i]
        if a in ("-r", "--room") and i + 1 < len(argv):
            i += 1; args["room"] = argv[i]
        elif a in ("-p", "--password") and i + 1 < len(argv):
            i += 1; args["password"] = argv[i]
        elif a in ("-n", "--name") and i + 1 < len(argv):
            i += 1; args["name"] = argv[i]
        elif a == "--server" and i + 1 < len(argv):
            i += 1; args["server"] = argv[i]
        elif a == "--limit" and i + 1 < len(argv):
            i += 1; args["limit"] = int(argv[i])
        else:
            args["_"].append(a)
        i += 1
    return args


# ========== HTTP Helper ==========

def http_request(method, url, body=None, headers=None):
    hdrs = dict(headers or {})
    data = None
    if body is not None:
        data = json.dumps(body).encode("utf-8")
        hdrs["Content-Type"] = "application/json"

    req = urllib.request.Request(url, data=data, headers=hdrs, method=method)
    try:
        with urllib.request.urlopen(req) as resp:
            raw = resp.read().decode("utf-8")
            try:
                return resp.status, json.loads(raw)
            except json.JSONDecodeError:
                return resp.status, {"ok": False, "error": raw or "invalid response"}
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8")
        try:
            return e.code, json.loads(raw)
        except json.JSONDecodeError:
            return e.code, {"ok": False, "error": raw or "HTTP error"}
    except urllib.error.URLError as e:
        raise ConnectionError(str(e.reason))


# ========== Resolve Server URL ==========

def resolve_server(args):
    server = args.get("server") or os.environ.get("SCR_SERVER") or load_config().get("server")
    if not server:
        return "http://localhost:3000"
    server = server.rstrip("/")
    if not server.startswith(("http://", "https://")):
        server = "http://" + server
    return server


# ========== Commands ==========

def cmd_login(args):
    room = args.get("room")
    user = args.get("name")
    if not room:
        sys.stderr.write("Error: -r <room> required\n")
        sys.exit(1)
    if not user:
        sys.stderr.write("Error: -n <name> required\n")
        sys.exit(1)

    server = resolve_server(args)
    body = {"room": room, "user": user}
    if args.get("password"):
        body["password"] = args["password"]

    try:
        status, data = http_request("POST", server + "/api/login", body)
        if data.get("ok"):
            save_config({
                "server": server,
                "token": data["token"],
                "room": data["room"],
                "roomName": data["roomName"],
                "user": data["user"],
            })
            print(f"Logged in as {data['user']} to {data['roomName']} on {server}")
        else:
            sys.stderr.write(f"Error: {data.get('error', 'login failed')}\n")
            sys.exit(1)
    except ConnectionError as e:
        sys.stderr.write(f"Error: Cannot connect to server at {server} - {e}\n")
        sys.exit(1)


def cmd_logout(_args):
    if os.path.exists(CONFIG_PATH):
        os.remove(CONFIG_PATH)
        print("Logged out. Session cleared.")
    else:
        print("Not logged in.")


def cmd_send(args):
    cfg = load_config()
    if not cfg.get("token"):
        sys.stderr.write("Not logged in. Run: ./scripts/scr login -r <room> -n <name>\n")
        sys.exit(1)

    text = " ".join(args["_"])
    if not text:
        sys.stderr.write("Error: message text required\n")
        sys.exit(1)

    server = resolve_server(args) or cfg.get("server")
    try:
        status, data = http_request("POST", server + "/api/send", {"text": text}, {
            "Authorization": "Bearer " + cfg["token"]
        })
        if data.get("ok"):
            print(f"ok {data['message']['id']}")
        elif status == 401:
            sys.stderr.write("Session expired. Please login again.\n")
            sys.exit(1)
        else:
            sys.stderr.write(f"Error: {data.get('error', 'send failed')}\n")
            sys.exit(1)
    except ConnectionError as e:
        sys.stderr.write(f"Error: Cannot connect to server at {server} - {e}\n")
        sys.exit(1)


def cmd_read(args):
    cfg = load_config()
    if not cfg.get("token"):
        sys.stderr.write("Not logged in. Run: ./scripts/scr login -r <room> -n <name>\n")
        sys.exit(1)

    after_id = args["_"][0] if args["_"] else None
    limit = args.get("limit", 50)
    server = resolve_server(args) or cfg.get("server")

    params = {"room": cfg["room"], "limit": str(limit)}
    if after_id:
        params["after"] = after_id
    url = server + "/messages?" + urllib.parse.urlencode(params)

    try:
        status, data = http_request("GET", url)
        if data.get("ok") and isinstance(data.get("messages"), list):
            msgs = data["messages"]
            if not msgs:
                print("(no new messages)")
                return
            for m in msgs:
                ts_str = m.get("ts", "")
                try:
                    ts = datetime.fromisoformat(ts_str.replace("Z", "+00:00"))
                    time_str = ts.strftime("%H:%M:%S")
                except Exception:
                    time_str = ts_str
                if m.get("type") == "file":
                    text = f"[file] {m.get('file', {}).get('name', '')}"
                else:
                    text = m.get("text", "")
                print(f"[{m['id']}] [{time_str}] <{m['user']}> {text}")
            print(f"LAST_ID:{msgs[-1]['id']}")
        else:
            sys.stderr.write(f"Error: {data.get('error', 'read failed')}\n")
            sys.exit(1)
    except ConnectionError as e:
        sys.stderr.write(f"Error: Cannot connect to server at {server} - {e}\n")
        sys.exit(1)


def show_help():
    print("""scr - SimpleChatRoom CLI

Usage:
  scr login -r <room> -n <name> [-p <password>] [--server <url>]
  scr logout
  scr send <message...>
  scr read [lastMsgId] [--limit N]

Options:
  -r, --room      Room name
  -n, --name      Username
  -p, --password  Room password (if required)
  --server        Server URL (default: http://localhost:3000)
  --limit         Max messages to read (default: 50)

Config: ~/.scr.json
Server priority: --server > $SCR_SERVER > config > localhost:3000""")


# ========== Main ==========

def main():
    argv = sys.argv[1:]
    if not argv:
        show_help()
        return

    cmd = argv[0]
    args = parse_args(argv[1:])

    commands = {
        "login": cmd_login,
        "logout": cmd_logout,
        "send": cmd_send,
        "read": cmd_read,
        "help": lambda _: show_help(),
        "--help": lambda _: show_help(),
        "-h": lambda _: show_help(),
    }

    func = commands.get(cmd)
    if func:
        func(args)
    else:
        sys.stderr.write(f"Unknown command: {cmd}\nRun: scr help\n")
        sys.exit(1)


if __name__ == "__main__":
    main()
